### PR TITLE
goto-synthesizer: Separate front-end for loop-contracts synthesizer

### PR DESCRIPTION
**Pull request by qinheping**
This introduces a new binary to handle all steps related to loop-contracts synthesis, which previously was part of goto-instrument. The main reason we want a separate front-end is to let one front-end tool to focus on one specific task. Also, there is no termination guarantee for the synthesizer which is not aligned with other functionality of goto-instrument.

No functionality is included in this PR yet. I will migrate the synthesizer into goto-synthesizer once this PR is merged.

### CODE REVIEW

File: src/Makefile
Comment (tautschnig) [inline src/Makefile]: Nitpick: please add a space before the backslash

File: src/goto-synthesizer/CMakeLists.txt
Comment (tautschnig) [inline src/goto-synthesizer/CMakeLists.txt:25]: Can you please also add a man page (in doc/man/) even when it will basically be empty at this point? And then please add an `install` rule for it in here plus amend `scripts/check_help.sh` to include the new binary.

File: src/goto-synthesizer/Makefile
Comment (tautschnig) [inline src/goto-synthesizer/Makefile]: Nit pick: no need for blank line

File: src/goto-synthesizer/goto_synthesizer_parse_options.cpp
Comment (tautschnig) [inline src/goto-synthesizer/goto_synthesizer_parse_options.cpp]: Hmmm ... it's 2022 after all :-)

File: src/goto-synthesizer/goto_synthesizer_parse_options.cpp
Comment (tautschnig) [inline src/goto-synthesizer/goto_synthesizer_parse_options.cpp]: We might be doing this elsewhere, but moving forward I don't think throwing exceptions in this piece of code is the right thing to do. You probably want to add a return value to this method, return `CPROVER_EXIT_USAGE_ERROR`, and have the caller `return` on any non-zero return value.

**Review by tautschnig — APPROVED** [review]
Can you please also amend `CODEOWNERS` to include a special rule for this new directory? (I think the rule should be the same as the existing one for the synthesizer.)

**Comment by tautschnig** [comment]
@kroening @peterschrammel Would welcome your thoughts here. @qinheping has implemented what I suggested in his other PR, and obviously I'm happy with that :-) Please speak up if you'd prefer not to add a separate binary.

File: src/goto-synthesizer/goto_synthesizer_parse_options.cpp
Comment (TGWDB) [inline src/goto-synthesizer/goto_synthesizer_parse_options.cpp]: Is this right?

File: src/goto-synthesizer/goto_synthesizer_parse_options.h
Comment (TGWDB) [inline src/goto-synthesizer/goto_synthesizer_parse_options.h]: Is this author right?

File: src/goto-synthesizer/goto_synthesizer_parse_options.cpp
Comment (TGWDB) [reply src/goto-synthesizer/goto_synthesizer_parse_options.cpp]: Or 2023 very soon!

File: src/goto-synthesizer/goto_synthesizer_parse_options.cpp
Comment (TGWDB) [reply src/goto-synthesizer/goto_synthesizer_parse_options.cpp]: Particularly in light of some modularity proposals, we'd like to move away from exceptions and aborts, so returning a useful value is better.

**Review by TGWDB — APPROVED** [review]
Looks ok with some minor fixes. Also happy to see a modular approach.

I think CODEOWNERS should be updated to reflect who should review PS here in future.

File: doc/man/goto-synthesizer.1
Comment (jimgrundy) [inline doc/man/goto-synthesizer.1]: "contracts" -> "contract's". Or:
"performs a given contracts synthesis" -> "performs synthesis for a given contract"
(if you are the kind of pedant who believe in using the possessive only for animate objects)

**Review by jimgrundy — APPROVED** [review]

**Comment by qinheping** [comment]
@NlightNFotis  The CI failure is caused by api codes as shown below. Could you please fix it?
```
/home/runner/work/cbmc/cbmc/src/libcprover-cpp/api.h should remove these lines:
- #include "options.h"  // lines 15-15
Unnecessary includes found. Use '// IWYU pragma: keep' to override this.
```

**Comment by thomasspriggs** [comment]
> @NlightNFotis The CI failure is caused by api codes as shown below. Could you please fix it?
> 
> ```
> /home/runner/work/cbmc/cbmc/src/libcprover-cpp/api.h should remove these lines:
> - #include "options.h"  // lines 15-15
> Unnecessary includes found. Use '// IWYU pragma: keep' to override this.
> ```

This issue in `develop` should now be fixed.

**Pull request merged**
